### PR TITLE
Fixing map_resave tool and add map_replace_image tool

### PR DESCRIPTION
**Pull request by sctt**
fixing map_resave tool by using the correct item and data sizes
(fixes [#843](https://github.com/ddnet/ddnet/issues/843))

### CODE REVIEW

File: src/tools/map_resave.cpp [hunk 1]
```diff
@@ -27,15 +27,15 @@ int main(int argc, const char **argv)
 	for(Index = 0; Index < DataFile.NumItems(); Index++)
 	{
 		pPtr = DataFile.GetItem(Index, &Type, &ID);
-		Size = DataFile.GetItemSize(Index);
+		Size = DataFile.GetRealItemSize(Index);
```
Comment (def-) [inline src/tools/map_resave.cpp]: Why is it not GetUncompressedSize here?

File: src/engine/shared/datafile.cpp [hunk 2]
```diff
@@ -362,6 +362,12 @@ int CDataFileReader::GetItemSize(int Index)
 	return m_pDataFile->m_Info.m_pItemOffsets[Index+1]-m_pDataFile->m_Info.m_pItemOffsets[Index];
 }
 
+int CDataFileReader::GetRealItemSize(int Index)
+{
+        if(!m_pDataFile) { return 0; }
```
Comment (def-) [inline src/engine/shared/datafile.cpp]: multiline please

**Comment by def-** [comment]
Oh, and you can add "(fixes #843)" to the commit message. Then github will automatically close #843 once this is merged.

File: src/engine/shared/datafile.cpp (hunk 2 above)
Comment (sctt) [reply src/engine/shared/datafile.cpp]: it's a copy and paste of the original GetItemSize...

File: src/tools/map_resave.cpp (hunk 1 above)
Comment (sctt) [reply src/tools/map_resave.cpp]: cause i didn't know that the mean of that fix was to uncompress item size, i'll fix.

**Comment by sctt** [comment]
i'm sorry for the multi commit, im not at my pc right now so im using github editor directly :/

**Comment by heinrich5991** [comment]
The `GetItemSize` function is of no use, could you rename it (if you want to keep it) and rename `GetUncompressedItemSize` to `GetItemSize`? That way you don't even need to touch the `src/tools/map_resave.cpp`.

**Comment by sctt** [comment]
i didn't do that cause i saw that in some points in the code the GetItemSize function is called, so i guessed it was needed somehow.

examples:
game/editor/io.cpp:615:   int ItemSize = DataFile.GetItemSize(Start) - 8;
tools/config_store.cpp:58:    int Size = Reader.GetItemSize(i) - sizeof(int) * 2;
game/server/gamecontext.cpp:2712:    int Size = Reader.GetItemSize(i) - sizeof(int) * 2;

are you sure we are not going to break anything if i just change his behavior?

**Comment by heinrich5991** [comment]
Ah, it looks like these function calls already work around the problem. Simply remove the `- 8` or `- 2 * sizeof(int)`. Good that you checked other function calls. ^^

**Comment by sctt** [comment]
oh right, but what about those points which don't fix the problem?
for example there is this:
`src/engine/shared/map.cpp:19:	virtual int GetItemSize(int Index) { return m_DataFile.GetItemSize(Index); }`
should i leave it like this?

**Comment by heinrich5991** [comment]
Yea, this function just forwards the implementation, that's fine.

**Comment by sctt** [comment]
ok i've fixed the other calls, i hope that's good

**Comment by sctt** [comment]
hello, i've also added the map_replace_image tool

**Comment by heinrich5991** [comment]
Is this more or less done now?

File: src/tools/map_replace_image.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+void* ReplaceImageItem(void *pItem, int Type, char *aImgName, char *aImgFile)
+{
+    if(Type != MAPITEMTYPE_IMAGE)
+        return pItem;
+
+    CMapItemImage *pImgItem = (CMapItemImage*) pItem;
+    char *pName = (char *)DataFile.GetData(pImgItem->m_ImageName);
+
+    if(str_comp(aImgName, pName))
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Just a suggestion, my personal preference, you don't have to add this: It doesn't hurt to add `!= 0` to make it clear that this is an inequality.

File: src/tools/map_replace_image.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        return pItem;
+
+    CImageInfo* pImgInfo = new CImageInfo;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: No need to allocate the `CImageInfo` on the heap. Remove the asterisk `*` and use the address-of operator `&` in the next line.

File: src/tools/map_replace_image.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    if(!LoadPNG(pImgInfo, aImgFile))
+        return pItem;
+
+    CMapItemImage *pNewImgItem = new CMapItemImage;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: This allocates and leaks the `CMapItemImage`. Either make it global (meh) or pass in a pointer. Ask me about this one if you don't know what to do. :)

File: src/tools/map_replace_image.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    pNewImgItem->m_Width  = pImgInfo->m_Width;
+    pNewImgItem->m_Height = pImgInfo->m_Height;
+    int PixelSize = pImgInfo->m_Format == CImageInfo::FORMAT_RGB ? 3 : 4;
+    
+    NewNameID = pImgItem->m_ImageName;
+    ExtractName(aImgFile, aNewName, sizeof(aNewName));
+    NewDataID = pImgItem->m_ImageData;
+    pNewData = pImgInfo->m_pData;
+    NewDataSize = pImgInfo->m_Width*pImgInfo->m_Height*PixelSize;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Spaces around operators please.

File: src/tools/map_replace_image.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    return (void *)pNewImgItem;
+}
+
+int main(int argc, const char **argv)
+{
+        IStorage *pStorage = CreateStorage("Teeworlds", IStorage::STORAGETYPE_BASIC, argc, argv);
+        int Index, ID = 0, Type = 0, Size;
+        void *pPtr;
+        char aFileName[1024], aImageName[1024], aImageFile[1024];
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Preferably declare one variable per line.

File: src/tools/map_replace_image.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    NewDataSize = pImgInfo->m_Width*pImgInfo->m_Height*PixelSize;
+
+    return (void *)pNewImgItem;
+}
+
+int main(int argc, const char **argv)
+{
+        IStorage *pStorage = CreateStorage("Teeworlds", IStorage::STORAGETYPE_BASIC, argc, argv);
+        int Index, ID = 0, Type = 0, Size;
+        void *pPtr;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: This is a very unspecific variable name. Maybe declare it further down, e.g. in the loops themselves so you can give it a more specific name.

File: src/tools/map_replace_image.cpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        char aFileName[1024], aImageName[1024], aImageFile[1024];
+
+        if(!pStorage || argc != 5) {
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Braces go on a new line in Teeworlds source code.

File: src/tools/map_replace_image.cpp [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+                return -1;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Maybe add some usage output (like the usage example you gave at the top of the file) if `argc != 5`.

File: src/tools/map_replace_image.cpp [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        str_format(aFileName, sizeof(aFileName), "%s", argv[2]);
+        str_format(aImageName, sizeof(aImageName), "%s", argv[3]);
+        str_format(aImageFile, sizeof(aImageFile), "%s", argv[4]);
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Simpler and more correct would be the following:
```
const char *pFileName = argv[2];
const char *pImageName = argv[3];
const char *pImageFile = argv[4];
```
This doesn't copy the strings around.

File: src/tools/map_replace_image.cpp [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+        if(!DataFile.Open(pStorage, argv[1], IStorage::TYPE_ALL)) {
+                return -1;
+        }
+        if(!df.Open(pStorage, aFileName)) {
+                return -1;
```
Comment (heinrich5991) [inline src/tools/map_replace_image.cpp]: Braces go on a new line. Maybe print an error message that says which operation failed.

File: src/tools/map_resave.cpp [hunk 13]
```diff
@@ -35,7 +35,7 @@ int main(int argc, const char **argv)
 	for(Index = 0; Index < DataFile.NumData(); Index++)
 	{
 		pPtr = DataFile.GetData(Index);
-		Size = DataFile.GetDataSize(Index);
+		Size = DataFile.GetUncompressedDataSize(Index);
```
Comment (heinrich5991) [inline src/tools/map_resave.cpp]: If, and *only* if, you're still motivated at this point, you could rename `GetDataSize` to `GetFileDataSize` and `GetUncompressedDataSize` to `GetDataSize`. Not necessary.

**Comment by sctt** [comment]
@heinrich5991 i made changes based on your suggestions and added more checks

**Pull request merged**